Add RSpec tests, fixtures and Travis CI support

ajlanghorn opened the pull request:
Adds a feature-complete Rakefile to support the running of syntax, spec, lint and metadata tests, as well as related fixtures, such as Gems. Also adds RSpec tests to check for desired behaviour, and finally a .travis.yml file to run tests on Travis CI.

Fixes #13.

ajlanghorn committed:
Add feature-complete Rakefile

This Rakefile is a copy of that at
gds-operations/puppet-module-skeleton@3e4e061
(https://github.com/gds-operations/puppet-module-skeleton/commit/3e4e061a7161dcb8b32fad60577752627da3259b).

It's more feature-complete than the original one, in that it sets up Puppet
Lint and Puppet Syntax tests, which can be used with the :test Rake task
with common-sense options.

ajlanghorn committed:
Add working Gitignore

Adds a sensible and working .gitignore based on
alphagov/puppet-module-skeleton.

ajlanghorn committed:
Add metadata-json-lint Gem

Allow for syntax-checking of the metadata.json file used to pass module
metadata in to the Puppet Forge using the metadata-json-lint Gem. The
Rakefile runs this as part of the "test" task.

ajlanghorn committed:
Add testing fixtures

Set-up dependency on puppetlabs/stdlib, and tell RSpec where to find the RCS
module.

ajlanghorn committed:
Add RSpec tests

Add tests to check for:
  - catalogue compilation errors
  - behaviour when no params passed, and defaults are used
  - behaviour when specific params are changed

ajlanghorn committed:
Add Travis CI support

Adds a .travis.yml file to test against Ruby 1.9.3 and Puppet 3.7.0 or
greater.

rjw1 commented:
did you mean to check in the Gemfile.lock since you added a gitignore entry for it.

ajlanghorn commented:
Good point, @rjw1. I'll get rid of that.

dcarley commented on spec/classes/rcs_spec.rb:
```diff
@@ -0,0 +1,81 @@
+
+  context 'sulogin is set to yes' do
+    let(:params) {{
+      :sulogin => 'yes',
```
Are lowercase strings a valid option?

ajlanghorn replied:
They appear to work in Vagrant, but as of https://github.com/alphagov/puppet-rcs/commit/195e40c375850d742937d81135b16fd673bdfaaf, they'll always be upcased because I call `upcase` on the string objects. That's tested in https://github.com/alphagov/puppet-rcs/commit/ff255235ea0a5090a25c21b7dfdebf29b808ea15.

dcarley replied:
Is that a branch from this branch that will form a later PR?

ajlanghorn replied:
Ish. It's a new branch from master, which will need rebasing on top of master before it becomes a later PR.

dcarley commented on spec/classes/rcs_spec.rb:
```diff
@@ -0,0 +1,81 @@
+    }}
+    it do
+      is_expected.to contain_file('/etc/default/rcS') \
+        .with_content(/TMPTIME=7/)
```
It might be worth anchoring these so that they only match a complete line, otherwise a stray comment would satisfy the test:

```
/^TMPTIME=7$/
```

ajlanghorn replied:
Huh. Anchoring makes the tests fail.

ajlanghorn replied:
Worked out why.

dcarley commented on spec/classes/rcs_spec.rb:
```diff
@@ -0,0 +1,81 @@
+        .with_content(/DELAYLOGIN=NO/)
+        .with_content(/UTC=YES/)
+        .with_content(/VERBOSE=NO/)
+        .with_content(/FSCKFIX=NO/)
```
I think it's often easier to read failures of these are broken out to separate `it{}` blocks so that they're tested and reported independently.

ajlanghorn replied:
That leads to a lot of duplication in the context block, which I thought looked messy?

ajlanghorn committed:
Anchor expected content regexes

The regexes given to the `with_content` matcher should be anchored to ensure
we don't pick up stray comments. Fixes.
